manual/manual/html_processing/js/search.js

```diff
@@ -0,0 +1,242 @@
+    if (err == -1 ||
+	(StartOnly && err != 0) ||
+	(EndOnly && err != s.length - sub.length)) {
+	err = MAX_ERROR;
+    } else {
+	if ( sub.includes("->") ) {
+	    err = Math.min(s.length - sub.length - err,1); // 0 or 1
+	    // err = 0 if the substring is right-aligned
+	} else {
+	    err = Math.min(err,1); // 0 or 1
+	    // err = 0 if the substring
+	}
+	err += Math.abs((s.length - sub.length) / s.length);}
+    return (err)
+    // between 0 and 2, except if MAX_ERROR
+}
+
+// Minimal substring error. In particular, it returns 0 if the string
+// 'sub' has an exact match with one of the strings in 'line'.
+function subMinError (sub, line) {
+    //let strings = [line[0], line[1], line[2]];
+    let errs = line.map(function (s) { return subError (sub, s); });
+    return Math.min(...errs); // destructuring assignment
+}
+
+
+function add (acc, a) {
+    return acc + a;
+}
+
+// for each sub we compute the minimal error within 'line', and then
+// take the average over all 'subs'. Thus it returns 0 if each sub has
+// an exact match with one of the strings in 'line'.
+function subsMinError (subs, line) {
+    let errs = subs.map(function (sub) { return subMinError (sub, line); });
+    return errs.reduce(add,0) / subs.length;
+}
+    
+
+function formatLine (line) {
+    let li = '<li>';
+    let html = '<code class="code"><a href="' + line[1] + '"><span class="constructor">' + line[0] + '</span></a>' +
+	'.' + '<a href="' + line[3] + '">' + line[2] + '</a></code>';
+    if (line.length > 5) {
+	if ( line[ERR_INDEX] == 0 ) {
+	    li = '<li class="match">';
+	}
+	html += (' : ' + line[SIG_INDEX] + '</pre>');
+	html = '<pre>' + html + line[DESCR_INDEX]; }
+    return (li + html + "</li>\n");
+}
+
+// The initial format of an entry of the GENERAL_INDEX array is
+// [ module, module_link,
+//   value, value_link,
+//   html_description, bare_description,
+//   html_signature, bare_signature ]
+
+// If includeDescr is false, the line is truncated to its first 4
+// elements.  When searching, the search error is added at the end of
+// each line.
+
+// In order to reduce the size of the index.js file, one could create
+// the bare_description on-the-fly using .textContent, see
+// https://stackoverflow.com/questions/28899298/extract-the-text-out-of-html-string-using-javascript,
+// but it would probably make searching slower (haven't tested).
+function mySearch (includeDescr) {
+    if (loadingIndex (includeDescr)) {
+	return;
+    }
+    let text = document.getElementById('api_search').value;
+    let results = [];
+    let html = "";
+    let count = 0;
+    let err_index = DESCR_INDEX;
+
+    if (text !== "") {
+	let words = [];
+	if ( includeDescr ) {
+	    err_index = ERR_INDEX;
+	    
+	    // Split text into an array of non-empty words:
+	    if ( text.includes("->") ) {
+		// this must be a type, so we don't split on single space
+		words = text.split("  ");
+	    } else {
+		words = text.split(" ");
+	    };
```

Comment:
Contextual separators sounds like a bad idea. I would propose to use `or` (since this is an OCaml keyword) or `|`.  Also is this feature used? Note that the search is currently inefficient for large number of alternatives (which seems fine?).

Reply:
what do you mean by inefficient (slow, or non-relevant results)?

I'm in favor of keeping the double space. It's not exactly a logical "or" because the ranking favors the match of both sides.
For instance if you want the signature 'int -> int' within the 'Int' module you can search "Int  int -> int"
Of course I have no idea whether this feature is used by others than me ;)

Reply:
With the current code, for k words the input will be scanned k times in `hasSubstrings` which is unnecessary. For large number of words, it is probably better to build an or-regexp.
Assigning two different meanings to space depending on the presence of arrows is not a good idea. Similarly, the difference between ` ` and `  ` is far too small. I am not opposed to the feature but it should use a non-ambiguous token (`;` for instance?).

Reply:
> With the current code, for k words the input will be scanned k times in `hasSubstrings` which is unnecessary.

really? I was assuming that `subs.findIndex` returns as soon as a match is found

Reply:
It does, but going from k to k/2 does not matter much for the asymptotic complexity.

Reply:
> Assigning two different meanings to space depending on the presence of arrows is not a good idea. Similarly, the difference between ` ` and ` ` is far too small. I am not opposed to the feature but it should use a non-ambiguous token (`;` for instance?).

I think it is important to keep a single space for the usual "google-like search" that everybody is now used to, so that if you type "print character" you get `Stdlib.print_char` as a result. However when searching for a type, necessarily the space acquires another meaning. But in most cases, my impression is that people will either search for a type or for keywords, but not both at the same time; then the problem disappears. If they really want to combine both, the fact that is becomes a bit tricky is not so scandalous, i think.

Reply:
The google version would be to use `"..."` for searching types. Your solution seems far too brittle and ad hoc. If you want to optimize for single type search, I would propose a more structured alternative: words are grouped with `"..."`, and the toplevel search is first considered as a single group and only broken down to words if there is no search result.

Reply:
yes I was thinking of using double quotes too; let's do this

Reply:
> It does, but going from k to k/2 does not matter much for the asymptotic complexity.

right, but I think we are talking here about k=2,3, or 4 maximum, don't you think so?

Reply:
I agree, thus the disclaimer that it is probably fine for the currently expected values of `k`.

Reply:
done. I also did some benchmarks to optimize search speed a bit